src/Effectful/HPQTypes.hs

```diff
@@ -65,80 +69,148 @@ instance DB :> es => PQ.MonadDB (Eff es) where
+    , PQ.dbRecordLastQuery = True
+    , PQ.dbQueryResult = Nothing
+    }
+
+handleAutoTransaction ::
+  PQ.TransactionSettings ->
+  (PQ.TransactionSettings -> m a -> m a) ->
+  m a ->
+  m a
+handleAutoTransaction transactionSettings withTransaction action =
+  -- We don't set tsAutoTransaction to False in the context of the action
+  -- because if the action calls commit inside, then with tsAutoTransaction
+  -- another transaction should be started automatically and if it's not set, it
+  -- won't happen (see source of the commit' function).  On the other hand,
+  -- withTransaction itself uses commit' and there we don't want to start
+  -- another transaction.
+  if PQ.tsAutoTransaction transactionSettings
+    then withTransaction (transactionSettings {PQ.tsAutoTransaction = False}) action
+    else action
+
+---------------------------------------------------
+-- Internal effect stack
+---------------------------------------------------
+
+-- | Newtype wrapper over the internal DB effect stack
+newtype DBEff es a = DBEff
+  { unDBEff :: Eff (State (DBState es) : es) a
+  }
+  deriving newtype (Functor, Applicative, Monad, MonadThrow, MonadCatch, MonadMask)
+
+-- | Internal state used to reinterpret the `DB` effect
+type DBState es = PQ.DBState (Eff es)
+
+-- Convenience `MonadIO` instance
+instance (IOE :> es) => MonadIO (DBEff es) where
+  liftIO b = DBEff $ liftIO b
+
+get :: DBEff es (DBState es)
+get = DBEff State.get
+
+put :: DBState es -> DBEff es ()
+put = DBEff . State.put
+
+modify :: (DBState es -> DBState es) -> DBEff es ()
+modify = DBEff . State.modify
+
+instance (IOE :> es, Error PQ.HPQTypesError :> es) => PQ.MonadDB (DBEff es) where
+  runQuery sql = do
+    dbState <- get
+    (result, dbState') <- liftIO $ PQ.runQueryIO sql dbState
+    put dbState'
+    pure result
+
+  getQueryResult =
+    get >>= \dbState -> pure $ PQ.dbQueryResult dbState
+
+  clearQueryResult =
+    modify $ \st -> st {PQ.dbQueryResult = Nothing}
+
+  getConnectionStats = do
+    dbState <- get
+    mconn <- liftIO . readMVar . PQ.unConnection $ PQ.dbConnection dbState
+    case mconn of
+      Nothing -> DBEff . throwError $ PQ.HPQTypesError "getConnectionStats: no connection"
+      Just cd -> pure $ PQ.cdStats cd
+
+  runPreparedQuery queryName sql = do
+    dbState <- get
+    (result, dbState') <- liftIO $ PQ.runPreparedQueryIO queryName sql dbState
+    put dbState'
+    pure result
+
+  getLastQuery = PQ.dbLastQuery <$> get
+
+  getTransactionSettings = PQ.dbTransactionSettings <$> get
+
+  setTransactionSettings settings = modify $ \st' ->
+    st' {PQ.dbTransactionSettings = settings}
+
+  withFrozenLastQuery action = do
+    let restoreRecordLastQuery st =
+          modify $ \st' ->
+            st' {PQ.dbRecordLastQuery = PQ.dbRecordLastQuery st}
+    bracket get restoreRecordLastQuery $ \st -> do
+      put st {PQ.dbRecordLastQuery = False}
+      action
+
+  withNewConnection action = DBEff $ do
```

Comment:
Nice!  Things get a bit complicated so it took me some time to understand what's going on :).  But it looks good to me.

In the process I realized we could (I think) also get away without `raiseWith` or even `bracket` and instead use `raise` and run separate internal state (as this is done in DBT I believe).  Something like:
```haskell
  withNewConnection action = do
    dbState0 <- get
    (result, dbState') <- DBEff . raise $ do
      PQ.withConnection (PQ.dbConnectionSource dbState0) $ \newConn -> do
        let transactionSettings = PQ.dbTransactionSettings dbState0
            dbState = mkDBState (PQ.dbConnectionSource dbState0) newConn transactionSettings
        runState dbState . unDBEff $ do
          handleAutoTransaction transactionSettings PQ.withTransaction' action
    result <$ put dbState'
```

I think it should do the same thing?  Maybe avoiding to run an internal state is better, though, I'm not sure :thinking:

Reply:
Oh this version actually doesn't work, it makes the withNewConnection test fail.  It looks like `action` still uses the external state, even though it's run within the context of a new, internal state.  Is that normal?  I will have to investigate that...

Reply:
Yes, that's normal (although somewhat subtle).

If you look at `runEffectDB` (we should rename it to `runDB` btw :slightly_smiling_face: ), the code that actually gets executed in tests when you call `withNewConnection` is the following:

```haskell
WithNewConnection (action :: Eff localEs b) ->
      unDBEff . PQ.withNewConnection . DBEff $
        localSeqUnlift env $
          \unlift -> unlift action
```

so the `action` is run using the provided `LocalEnv` (which really is just an opaque wrapper for `Env` that's in use at the point where `withNewConnection` was called in tests). This `LocalEnv` somewhere inside has a handler for `DB` and the handler holds a reference to `Env` of the handler that has the original `State` created by `runWithState`.

Now, when you call `raise` in the handler, **a new branch** of `Env` that can be modified independently of the original `Env` of the handler is created (it has to be independent since if it wasn't, you could call `raise` and then overwrite the last effect on the stack with whatever you want). The `LocalEnv` doesn't see that independent branch anymore, hence it won't see the `State` that is created after the `raise` in that independent branch.

That's a minus of using the newtype and intermediate instance of `DBEff` - this is not immediately obvious at all until you mentally inline the definition of `withNewConnection` from the `DBEff` instance into the interpreter.

Does that make sense?

In fact, in the beginning I did what you did and was confused for a moment why it doesn't work, until I realized the above :slightly_smiling_face:

Reply:
> If you look at runEffectDB (we should rename it to runDB btw slightly_smiling_face ), the code that actually gets executed in tests when you call withNewConnection is the following:

Ah yes, that's what I was missing, I guess that makes sense :slightly_smiling_face:.

Still, this entire combination of `raise` and `runState` with unlifted action is somewhat confusing to me.  Since the `dbState` provided to the internal `runState` is irrelevant (because `action` uses the state provided in `LocalEnv`), as in the following snippet:
```haskell
  ...
  raise $ do
    ...
    runState dbState $ localSeqUnlift env $ \unlift -> unlift action
```
one could expect the following to have the same behavior:
```haskell
  ...
  raise $ do
    ...
    runState () $ localSeqUnlift env $ \unlift -> unlift action
```
but the second version doesn't compile - the internal state might be irrelevant, but it must have the same type as the original state :exploding_head:.

But I guess I will get used to that :slightly_smiling_face:.

Reply:
> Still, this entire combination of `raise` and `runState` with unlifted action is somewhat confusing to me. Since the `dbState` provided to the internal `runState` is irrelevant (because `action` uses the state provided in `LocalEnv`), as in the following snippet:
> 
> ```haskell
>   ...
>   raise $ do
>     ...
>     runState dbState $ localSeqUnlift env $ \unlift -> unlift action
> ```
> 
> one could expect the following to have the same behavior:
> 
> ```haskell
>   ...
>   raise $ do
>     ...
>     runState () $ localSeqUnlift env $ \unlift -> unlift action
> ```
> 
> but the second version doesn't compile - the internal state might be irrelevant, but it must have the same type as the original state exploding_head.

It doesn't, but for unrelated reasons. It doesn't compile because `SuffixOf` constraint won't let it. This was supposed to prevent doing a few nasty things, but it turns out it's too restrictive, which causes confusion (as you found out) and also it doesn't really behave as advertised because type families are weird. I fixed it in https://github.com/haskell-effectful/effectful/pull/74 (and documented the "nasty things" in the haddock). After this is merged, your second version will work as  well as this one:

```haskell
  ...
  raise $ do
    ...
    localSeqUnlift env $ \unlift -> unlift action
```

Reply:
Great, this is much more predictable indeed, thanks!